### PR TITLE
Adding functionality to create a manifest list from a list of built images

### PR DESCRIPTION
**Pull request by louismurerwa**
This PR adds the ability to build a manifest list from a list of images
This PR also adds the additional testing required to ensure the functionality of the added features.

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListStep.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import com.google.cloud.tools.jib.json.JsonTemplate;
+import com.google.common.io.ByteStreams;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+/** Creates a manifest list. */
+class BuildManifestListStep implements Callable<ManifestTemplate> {
+
+  private static final String DESCRIPTION = "Creating a manifest list";
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListStep.java:39]: super-nit: we've been saying "Building ..." for others. Let's go for consistency.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListStep.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  @Override
+  public ManifestTemplate call() throws IOException {
+
+    EventHandlers eventHandlers = buildContext.getEventHandlers();
+    try (TimerEventDispatcher ignored = new TimerEventDispatcher(eventHandlers, DESCRIPTION);
+        ProgressEventDispatcher ignored2 =
+            progressEventDispatcherFactory.create("creating a manifest list", 1)) {
+      eventHandlers.dispatch(LogEvent.info("Creating a manifest list"));
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListStep.java:61]: "building ..."

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListStep.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      JsonTemplate containerConfiguration =
+          new ImageToJsonTranslator(builtImages.get(0)).getContainerConfiguration();
+      BlobDescriptor configDescriptor =
+          Blobs.from(containerConfiguration).writeTo(ByteStreams.nullOutputStream());
+
+      return new ImageToJsonTranslator(builtImages.get(0))
+          .getManifestTemplate(buildContext.getTargetFormat(), configDescriptor);
+    }
+
+    return new ManifestListGenerator(this.buildContext, this.builtImages).getManifestListTemplate();
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListStep.java:73]: Put everything into the try-resources block.

```java
  try (TimerEventDispatcher ...) {
    ...
    return new ManifestListGenator(...);
  }
```
And I think this can be simplified by
```java

Put everything into the try-resources block.

```java
  try (TimerEventDispatcher ...) {
    ...
    return new ManifestListGenator(...);
  }
```

And actually, I think we can use
```java
BlobDescriptor configDescriptor = Digests.computeDigest(containerConfig);
```
instead of `Blobs.from(...).writeTo()`. I am not sure why we used it that way in other places.

And I think we can simply this to
```java
    if (builtImages.size() == 1) {
      ImageToJsonTranslator imageTranslator = new ImageToJsonTranslator(builtImages.get(0));
      BlobDescriptor configDescriptor =
          Digests.computeDigest(imageTranslator.getContainerConfiguration());
      return imageTranslator.getManifestTemplate(buildContext.getTargetFormat(), configDescriptor);
    }

    return new ManifestListGenerator(...);
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java [hunk 4]
```diff
@@ -423,6 +426,21 @@ private void buildImages() {
             });
   }
 
+  private void buildManifestList() {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:429]: buildManifestListOrSingleManifest

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java [hunk 5]
```diff
@@ -423,6 +426,21 @@ private void buildImages() {
...
+
+    results.manifestListOrSingleManifest =
+        executorService.submit(
+            () -> {
+              List<Future<Image>> builtImages = new ArrayList<>();
+              builtImages.addAll(results.builtImagesAndBaseImages.get().keySet());
+              return new BuildManifestListStep(
+                      buildContext, childProgressDispatcherFactory, realizeFutures(builtImages))
+                  .call();
+            });
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:441]: This is confusing

```suggestion
            () ->
              new BuildManifestListStep(
                      buildContext,
                      childProgressDispatcherFactory,
                      realizeFutures(results.builtImagesAndBaseImages.get().keySet()))
                  .call());
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import com.google.cloud.tools.jib.blob.Blobs;
+import com.google.cloud.tools.jib.configuration.BuildContext;
+import com.google.cloud.tools.jib.image.Image;
+import com.google.cloud.tools.jib.image.json.V22ManifestListTemplate.ManifestDescriptorTemplate;
+import com.google.cloud.tools.jib.json.JsonTemplate;
+import com.google.common.io.ByteStreams;
+import java.io.IOException;
+import java.util.List;
+
+/** Translates a list of {@link Image} into a manifest list. */
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:29]:
```suggestion
/** Generator a manifest list for {@link Image}s. */
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+public class ManifestListGenerator {
+
+  private final BuildContext buildContext;
+  private final List<Image> builtImages;
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:33]: `ManifestListGenerator` doesn't and shouldn't care what the images are. It's an isolated class.

```suggestion
  private final List<Image> images;
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  public ManifestListGenerator(BuildContext buildContext, List<Image> builtImages) {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:35]: From the API design perspective, it shouldn't know anything about `BuildContext`. It's an isolated class and shouldn't be coupled.

```suggestion
  public ManifestListGenerator(... imageFormat, List<Image> images) {
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    this.buildContext = buildContext;
+    this.builtImages = builtImages;
+  }
+
+  /**
+   * Translates a list of {@link Image} into a manifestTemplate.
+   *
+   * @return ManifestTemplate a JSON representation of list of {@link Image}.
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:43]: No period at the end
```suggestion
   * @return a manifest list JSON
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  private final BuildContext buildContext;
+  private final List<Image> builtImages;
+
+  public ManifestListGenerator(BuildContext buildContext, List<Image> builtImages) {
+    this.buildContext = buildContext;
+    this.builtImages = builtImages;
+  }
+
+  /**
+   * Translates a list of {@link Image} into a manifestTemplate.
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:41]: `Generates a manifest list JSON for the given {@link Image}s.`

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  public ManifestTemplate getManifestListTemplate() {
+
+    try {
+      V22ManifestListTemplate manifestList = new V22ManifestListTemplate();
+
+      for (Image builtImage : builtImages) {
+        JsonTemplate containerConfiguration =
+            new ImageToJsonTranslator(builtImage).getContainerConfiguration();
+        BlobDescriptor configDescriptor =
+            Blobs.from(containerConfiguration).writeTo(ByteStreams.nullOutputStream());
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:54]: ditto

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        ManifestDescriptorTemplate manifest = new ManifestDescriptorTemplate();
+        manifest.setMediaType(manifestTemplate.getManifestMediaType());
+        manifest.setSize(manifestDescriptor.getSize());
+        manifest.setDigest(manifestDescriptor.getDigest().toString());
+        manifest.setPlatform(builtImage.getArchitecture(), builtImage.getOs());
+        manifestList.addManifest(manifest);
+      }
+      return manifestList;
+
+    } catch (IOException ex) {
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:71]: I need to think about this. Probably not the right way.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 13]
```diff
@@ -68,13 +70,32 @@
...
+  }
+
+  public String getMediaType() {
+    return mediaType;
   }
 
   @Nullable private List<ManifestDescriptorTemplate> manifests;
 
+  /**
+   * Adds a manifest to the manifestList.
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java:88]:
```suggestion
   * Adds a manifest.
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java [hunk 14]
```diff
@@ -68,13 +70,32 @@
...
+   *
+   * @param manifest a JSON representation of an {@link Image}
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/V22ManifestListTemplate.java:90]: This class shouldn't mention `Image`. The class is totally unrelated. I think we can just go with

```suggestion
   * @param manifest a manifest descriptor
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java (hunk 5 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:441]: We need to convert the `keySet()` to an ArrayList before using the `realizeFutures` .

`   new BuildManifestListOrSingleManifestStep(
                        buildContext,
                        childProgressDispatcherFactory,
                        realizeFutures(
                            new ArrayList<>(results.builtImagesAndBaseImages.get().keySet())))
                    .call());

`
This might be better

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:441]: Aha, I think we should change `realizeFutures()` instead.
```java
  private static <E> List<E> realizeFutures(List<Future<E>> futures)
```
```java
  private static <E> List<E> realizeFutures(Collection<Future<E>> futures)
```
`Set`, `List`, `Vector`, etc. all implement `Collection`.

Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:441]: Sounds Good I will do a small PR to change `realizeFutures`

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:441]: It will be a one-line change. If you prefer, you can do it here.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java (hunk 12 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:66]: I think this method should not catch this exception. Let's just throw it.

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStep.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        ImageToJsonTranslator imageTranslator = new ImageToJsonTranslator(builtImages.get(0));
+        BlobDescriptor configDescriptor =
+            Digests.computeDigest(imageTranslator.getContainerConfiguration());
+        return imageTranslator.getManifestTemplate(
+            buildContext.getTargetFormat(), configDescriptor);
+      }
+
+      eventHandlers.dispatch(LogEvent.info("Building a manifest list"));
+      return new ManifestListGenerator(this.builtImages)
+          .getManifestListTemplate(this.buildContext.getTargetFormat());
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStep.java:72]:
```suggestion
      return new ManifestListGenerator(builtImages)
          .getManifestListTemplate(buildContext.getTargetFormat());
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java [hunk 16]
```diff
@@ -206,6 +209,7 @@ public BuildResult run() throws ExecutionException, InterruptedException {
       rootProgressDispatcher = progressEventDispatcher;
 
       stepsToRun.forEach(Runnable::run);
+      realizeFutures(results.buildResults.get());
```
[the remark this answers is not in the document]
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java:212]: Sounds let me remove it.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+     * },
+     * "layers":[
+     * {
+     * "mediaType":"application/vnd.docker.image.rootfs.diff.tar.gzip",
+     * "size":0
+     * }
+     * ]
+     * }
+     * }</pre>
+     */
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java:91]: Javadoc comments (starting with `/**` using double asterisks) are basically for machine processing to generate a nice-looking manual for external users using a library. This is an internal code comment, so we should use `// code comment`. And no HTML formatting, and no need to follow the Javadoc convention. I also suggest proper indentation for readability.

```
// Expected manifest JSON:
// {
//   "schemaVersion":2,
//   "mediaType":"application/vnd.docker.distribution.manifest.v2+json",
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+  public void setUp() {
+    Mockito.when(buildContext.getEventHandlers()).thenReturn(EventHandlers.NONE);
+    Mockito.doReturn(V22ManifestTemplate.class).when(buildContext).getTargetFormat();
+    Mockito.when(layer.getBlobDescriptor()).thenReturn(new BlobDescriptor(0, null));
+
+    image1 =
+        Image.builder(V22ManifestTemplate.class)
+            .setArchitecture("amd64")
+            .setOs("linux")
+            .addLayer(new PreparedLayer.Builder(layer).setName("resources").build())
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java:59]:
```suggestion
            .addLayer(layer)
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+            .build();
+    image2 =
+        Image.builder(V22ManifestTemplate.class)
+            .setArchitecture("arm64")
+            .setOs("ubuntu")
+            .addLayer(new PreparedLayer.Builder(layer).setName("classes").build())
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java:65]:
```suggestion
            .addLayer(layer)
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+   * Generates a manifest list JSON for the given {@link Image}s.
+   *
+   * @param <T> child type of {@link BuildableManifestTemplate}
+   * @param manifestTemplateClass the JSON template to translate the image to
+   * @return a manifest list JSON
+   */
+  public <T extends BuildableManifestTemplate> ManifestTemplate getManifestListTemplate(
+      Class<T> manifestTemplateClass) {
+    try {
+      V22ManifestListTemplate manifestList = new V22ManifestListTemplate();
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:45]: I see we always create a Docker manifest list, which is another limitation of multi-platform image building. Note that there are 2 avenues for considering OCI vs Docker format.
- Base image: we take care of this automatically, and the format is transparent to the user.
- Target image(s): user-configurable with `<format>OCI` or `<format>Docker`.
That is, Jib will be broken when `<format>OCI` is specified (because we will build OCI manifests while a Docker manifest). We should add another item to fix. Please add this to your tracking list.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+    image1 =
+        Image.builder(V22ManifestTemplate.class)
+            .setArchitecture("amd64")
+            .setOs("linux")
+            .addLayer(new PreparedLayer.Builder(layer).setName("resources").build())
+            .build();
+    image2 =
+        Image.builder(V22ManifestTemplate.class)
+            .setArchitecture("arm64")
+            .setOs("ubuntu")
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java:64]: nit: let's say `"windows"` to eliminate any potential misunderstanding. (Ubuntu is linux).

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java (hunk 20 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:45]: Note that support OCI for target images is much more important: most registries can just convert OCI to Docker on-the-fly, so not accepting an OCI base image manifest isn't really a problem in practice. OTOH, specifying `<format>OCI` will outright breaking Jib image building (in the multi-platform case).

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java (hunk 17 above)
Comment (louismurerwa) [reply jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java:91]: Yes was the issue I was asking about in the meeting.I'l remove the java doc format and use code comments instead.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java (hunk 20 above)
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:45]: Added

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+            .addLayer(layer)
+            .build();
+  }
+
+  @Test
+  public void testCall_singleManifest() throws IOException {
+
+    // Expected manifest JSON
+    //  {
+    // "schemaVersion":2,
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java:74]:
```suggestion
    //  "schemaVersion":2,
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+    Assert.assertEquals(0, manifest.getLayers().get(0).getSize());
+    Assert.assertEquals(158, manifest.getContainerConfiguration().getSize());
+  }
+
+  @Test
+  public void testCall_manifestList() throws IOException {
+
+    // Expected Manifest List JSON
+    //  {
+    // "schemaVersion":2,
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/steps/BuildManifestListOrSingleManifestStepTest.java:111]:
```suggestion
    //  "schemaVersion":2,
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+    testGetManifestListTemplate(V22ManifestTemplate.class);
+  }
+
+  /** Tests translation of image to {@link BuildableManifestTemplate}. */
+  private <T extends BuildableManifestTemplate> void testGetManifestListTemplate(
+      Class<T> manifestTemplateClass) throws IOException {
+
+    // Expected Manifest List JSON
+    //  {
+    // "schemaVersion":2,
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java:50]:
```suggestion
    //  "schemaVersion":2,
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+import org.junit.Test;
+
+/** Tests for {@link ManifestListGenerator}. */
+public class ManifestListGeneratorTest {
+
+  private Image image1;
+  private Image image2;
+  private ManifestListGenerator manifestListGenerator;
+
+  private void setUp(Class<? extends BuildableManifestTemplate> imageFormat) {
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java:32]:
```suggestion
  @Before
  private void setUp() {
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  @Test
+  public void testGetManifest_v22() throws IOException {
+    setUp(V22ManifestTemplate.class);
+    testGetManifestListTemplate(V22ManifestTemplate.class);
+  }
+
+  /** Tests translation of image to {@link BuildableManifestTemplate}. */
+  private <T extends BuildableManifestTemplate> void testGetManifestListTemplate(
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java:45]:
```suggestion
  @Test
  public void testGetManifestListTemplate(
```

File: jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+    //    {
+    //      "mediaType":"application/vnd.docker.distribution.manifest.v2+json",
+    //      "digest":"sha256:51038a7a91c0e8f747e05dd84c3b0393a7016ec312ce384fc945356778497ae3",
+    //      "size":264,
+    //      "platform":{
+    //        "architecture":"arm64",
+    //        "os":"windows"
+    //      }
+    //    }
+    //  ]
```
Comment (chanseokoh) [inline jib-core/src/test/java/com/google/cloud/tools/jib/image/json/ManifestListGeneratorTest.java:71]:
```suggestion
    //   ]
```

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+   * Generates a manifest list JSON for the given {@link Image}s.
+   *
+   * @param <T> child type of {@link BuildableManifestTemplate}
+   * @param manifestTemplateClass the JSON template to translate the image to
+   * @return a manifest list JSON
+   * @throws IOException IOException
+   */
+  public <T extends BuildableManifestTemplate> ManifestTemplate getManifestListTemplate(
+      Class<T> manifestTemplateClass) throws IOException {
+
```
[the remark this answers is not in the document]
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:45]: I usually put for readability but I'l make sure to remove the spaces now

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java [hunk 29]
```diff
@@ -481,7 +498,6 @@ private void checkImageInTargetRegistry() {
     results.manifestCheckResult =
         executorService.submit(
             () -> {
-              Verify.verify(results.builtImagesAndBaseImages.get().size() == 1);
```
[the remark this answers is not in the document]
Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java]: Why do we need to keep this ?

[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java]: Our code is broken at HEAD if `results.builtImagesAndBaseImages.get().size() != 1`, so I think keeping as many checks as possible helps while we fix our code.

File: jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+   * @param manifestTemplateClass the JSON template to translate the image to
+   * @return a manifest list JSON
+   * @throws IOException if generating a manifest list fails due to an I/O error when computing
+   *     digests
+   */
+  public <T extends BuildableManifestTemplate> ManifestTemplate getManifestListTemplate(
+      Class<T> manifestTemplateClass) throws IOException {
+    Preconditions.checkArgument(
+        manifestTemplateClass == V22ManifestTemplate.class,
+        "Build an OCI image index is not yet supported");
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:49]: Oh, we should consider the boundary condition when `images` is empty. I think it makes sense to just fail

```
Preconditions.checkState(!images.isEmpty(), "no images given");
```

and add a test for this. (It's important to test boundary conditions.) It also makes sense to add the same test in `BuildManifestListOrSingleManifestStepTest` to see if the exception is properly propagated to that level too.

Comment (louismurerwa) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:49]: That makes sense , I was thinking about these cases from the `StepsRunner` standpoint. Since steps runner always ensures the `!images.isEmpty()`  we should be good but I am now realizing that these libraries can be called from anywhere in Jib so we actually need to check the parameters also inside the libraries themselves.

Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/image/json/ManifestListGenerator.java:49]: Exactly. The class is independent of `StepsRunner` (or anything else unrelated). And as soon as we have the class, other parts of the Jib repo will pick it up and start using it. The implementation should be self-sufficient and error-tolerant.
